### ISSUE TITLE
FCS AST and TAST should contain trivia

### ISSUE THREAD

**Issue by forki** [fsharp/fslang-suggestions#595]
I propose we the FSharp.Compiler.Service collects trivia like whitespace and returns it to clients.
This way it would probably be easier to develop formatter for ionide and Visual Studio.

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

**Comment by 7sharp9** [comment]
I raised this in FCS a while back there will be a closed issue somewhere ...

**Comment by ErikSchierboom** [comment]
This would be a great feature to have. Makes formatting so much easier.

**Comment by rmunn** [comment]
@7sharp9 Is https://github.com/fsharp/FSharp.Compiler.Service/issues/302 the closed issue you're referring to?

**Comment by rmunn** [comment]
https://github.com/fsharp/FSharp.Compiler.Service/issues/604 also looks relevant.

**Comment by alfonsogarciacaro** [comment]
If this is implemented, could you please make it optional? ;) BTW, comments are already accessible from the TAST (`FSharpMemberOrFunctionOrValue.XmlDoc`, see https://github.com/fable-compiler/Fable/issues/752) though you don't get the ranges.

**Comment by forki** [comment]
I'm not talking about xml docs. I'm talking about all comments in normal code (within methods). What I'd like to do is doing .ToString() on the AST and get the source code back.

**Comment by dsyme** [comment]
I'm not actually sure how to do this given the parser/lexer we use.  Perhaps it's easy, but as mentioned above we've had bugs integrating the XML comments - which we try to "grab" in pars.fsy - and fully accurate integration of trivia seems harder.  

Perhaps there's a simpler "second phase" approach to this - though it might end up similar to that used by Fantomas.

My understanding is that, after transformation, Fantomas re-tokenizes the original source (with whitespace/comments being reported) and tries to fold back comments into the transformed source.  It might be possible to do that immediately after parsing, augmenting the AST by filling in gaps left for syntax trivia.

**Comment by cartermp** [comment]
@dsyme Just as an aside, the Roslyn team refers to trivia jokingly as the worst-named concept they have to deal with, because it sounds so much like "trivial" when it's anything but that 😄 

I do very much like this suggestion, especially with the likely addition of fantomas to the language service.

**Comment by dsyme** [comment]
> I do very much like this suggestion, especially with the likely addition of fantomas to the language service.

In a way I would want Fantomas (and in particular its tests) integrated first, so when we implemented the trivia we would have a consumer of the trivia information - and tests for it - in the code base itself.   No one is going to update Fantomas to use the trivia "from the outside", but they may update it as part of implementing the trivia feature itself.

**Comment by forki** [comment]
Yes absolutely agree. Just adding trivia without dogfood ingredients it
would probably be very hard.

Am 05.09.2017 5:20 nachm. schrieb "Don Syme" <notifications@github.com>:

**Comment by dsyme** [comment]
@forki Some thoughts on this, as part of thinking how we might change things in FCS to help systematically improve the correctness of the proposed code formater.

First, in a rudimentary sense "trivia" is available for the FCS token stream - e.g. WHITESPACE tokens are reported as part of FSharpSourceTokenizer, because internally [``skip`` is set to false](https://github.com/Microsoft/visualfsharp/blob/44fa027b308681a1b78a089e44fa1ab35ff77b41/src/fsharp/vs/ServiceLexing.fs#L497)

Fantomas uses this already - first it parses the original code to an FCS syntax tree, reformats the syntax tree, prints it to the (commentless) result code. It then re-tokenizes both the result code and the original cpde, revealing all whitespace and comments and directives in the original, and then it tries to re-integrate those into the token stream of the output.

The problem is that FCS syntax trees don't integrate any real token information - there are just source range markers, and even these are occasionally inaccurate.  There are no tokens, and certainly no trivia tokens.

The thing I don't yet understand is if/how Roslyn maintains the correspondence between syntax tree nodes and actual tokens, and whether it would be useful to do that in FCS.

**Comment by TIHan** [comment]
Been looking at how this works in Roslyn, here is what I think is happening:

In C#'s lexer for getting tokens, the first thing is tries to do is lex all trivia leading up to a token, then any trailing trivia after that. Leading trivia could be comments, documentation, etc. and trailing could be whitespace. Trivia themselves are not tokens, they are actually their own syntax nodes that are unconnected to the language syntax nodes, called `SyntaxTrivia`, which are passed to the creation of a token.
```csharp
internal static SyntaxToken Create(SyntaxKind kind, GreenNode leading, GreenNode trailing)
```
This means, tokens are the ones that hold onto leading and trailing trivia.

For C#, this is really an internal implementation detail. What really matters for them is the red tree/red nodes as those are the public facades. A public syntax red node can be asked for leading trivia, and the first thing it does is grab the first token and ask it for any leading trivia.

For F#'s case, we have to do something different than this because F# throws away the tokens. Also, its syntax tree is structured differently; there is no concept of a "node" for F#'s parse tree. The parse tree is really just different forms of data the further you go down. The reason it's built like that is because it clearly defines the structure of what syntax is possible at different layers and context.

So what can we do? I think the best approach is to not heavily modify the existing parse tree that F# has. Instead, we build a second parse tree (lazily) with a concept of a "node" and we keep the tokens around. The relationship between the nodes and tokens can be determined by their ranges. This is how I approached it when building the compilation prototype; it seemed to work fairly well but with very limited use and testing.

**Comment by 7sharp9** [comment]
The main trivia I was referring to was code comments.
